programs/margin-account/src/lib.rs

```diff
@@ -70,22 +193,81 @@ pub struct Withdraw<'info> {
     token_program: AccountInfo<'info>,
 }
 
+// OpenPositionAMM takes the tokens that are in the margin account and executes a trade with them.
 #[derive(Accounts)]
-pub struct OpenPosition<'info> {
-    // TODO
+pub struct OpenPositionAMM<'info> {
     #[account(signer)]
-    authority: AccountInfo<'info>,
+    trader: AccountInfo<'info>,
+    /// accounts needed to call
+    swap_program: AccountInfo<'info>,
+    swap_info: AccountInfo<'info>,
+    swap_authority: AccountInfo<'info>,
+    #[account(mut)]
+    source: AccountInfo<'info>,
+    #[account(mut)]
+    swap_source: AccountInfo<'info>,
+    #[account(mut)]
+    swap_dest: AccountInfo<'info>,
+    #[account(mut)]
+    pool_mint: AccountInfo<'info>,
+    #[account(mut)]
+    pool_fee: AccountInfo<'info>,
+    host_fee: AccountInfo<'info>,
+    /// accounts needed to access funds from token vault
+    #[account(mut, has_one = trader, has_one = loaned_vault)]
+    margin_account: ProgramAccount<'info, MarginAccount>,
+    #[account(mut)]
+    loaned_vault: CpiAccount<'info, TokenAccount>,
+    #[account(mut)]
+    collateral_vault: CpiAccount<'info, TokenAccount>,
+    #[account(seeds = [margin_account.to_account_info().key.as_ref(), &[margin_account.nonce]])]
+    vault_signer: AccountInfo<'info>,
+
+    #[account("token_program.key == &token::ID")]
+    token_program: AccountInfo<'info>,
 }
 
+// ClosePositionAMM call an amm to close the entire position or only enough to repay the loan, if in profit
 #[derive(Accounts)]
-pub struct ClosePosition<'info> {
-    // TODO
+pub struct ClosePositionAMM<'info> {
     #[account(signer)]
-    authority: AccountInfo<'info>,
+    trader: AccountInfo<'info>,
+    /// accounts needed to call
+    swap_program: AccountInfo<'info>,
+    swap_info: AccountInfo<'info>,
+    swap_authority: AccountInfo<'info>,
+    #[account(mut)]
+    source: AccountInfo<'info>,
+    #[account(mut)]
+    swap_source: AccountInfo<'info>,
+    #[account(mut)]
+    swap_dest: AccountInfo<'info>,
+    #[account(mut)]
+    pool_mint: AccountInfo<'info>,
+    #[account(mut)]
+    pool_fee: AccountInfo<'info>,
+    host_fee: AccountInfo<'info>,
+    /// accounts needed to access funds from token vault
+    #[account(mut, has_one = trader, has_one = loaned_vault)]
+    margin_account: ProgramAccount<'info, MarginAccount>,
+    #[account(mut)]
+    loaned_vault: CpiAccount<'info, TokenAccount>,
+    #[account(mut)]
+    collateral_vault: CpiAccount<'info, TokenAccount>,
+    #[account(seeds = [margin_account.to_account_info().key.as_ref(), &[margin_account.nonce]])]
+    vault_signer: AccountInfo<'info>,
+
+    #[account("token_program.key == &token::ID")]
+    token_program: AccountInfo<'info>,
 }
 
 //? Possibly add cancel position, if it cannot be combined with close.
 
+#[derive(Accounts)]
+pub struct Repay<'info> {
+    // TODO
+    authority: AccountInfo<'info>,
+}
```

Comment:
why is repay a call into the margin account? Should this not just be a CPI call within the close/liquidate? Or are you worried of hitting a VM cap?